File: 07/bags.py
from collections import defaultdict
import re

color_contents = defaultdict(list)
color_ancestors = defaultdict(list)

for line in open('input.txt').readlines():
    color, rest = line.split(' bags contain ')
    if rest == "no other bags.":
        continue
    bag_amounts = re.findall('(\d+) ([^\.,]+) bags?', rest)
    color_contents[color] = bag_amounts
    for _, name in bag_amounts:
        color_ancestors[name].append(color)

# in how many different bags we can put a shiny gold bag
stack = ['shiny gold']
ancestors = set()
while stack:
    color = stack.pop(0)
    for ancestor in color_ancestors[color]:
        ancestors.add(ancestor)
        stack.append(ancestor)

print(len(set(ancestors)))

# how many other bags fit in a shiny gold bag
def count_bags_inside(color):
    contents = color_contents[color]
    return 1 + sum([int(n)*count_bags_inside(color) for (n, color) in contents])

print(count_bags_inside('shiny gold')-1)

File: 07/bags.go
package main

import (
	"fmt"
	"io/ioutil"
	"regexp"
	"strconv"
	"strings"
)

var colorContents = map[string][][]string{}
var colorAncestors = map[string][]string{}

func countBagsInside(color string) int {
	count := 1
	for _, amountOfColor := range colorContents[color] {
		n, _ := strconv.Atoi(amountOfColor[0])
		color := amountOfColor[1]
		count += n * countBagsInside(color)
	}
	return count
}

func main() {
	data, _ := ioutil.ReadFile("input.txt")
	text := strings.TrimSpace(string(data))
	re := regexp.MustCompile(`(\d+) ([^\.,]+) bags?`)
	for _, rule := range strings.Split(text, "\n") {
		s := strings.Split(rule, " bags contain ")
		color, rest := s[0], s[1]
		if rest == "no other bags." {
			continue
		}
		bagAmounts := re.FindAllStringSubmatch(rest, -1) // returns triplets of [matched_str, group_1, group_2]
		for _, bagAmount := range bagAmounts {
			colorContents[color] = append(colorContents[color], []string{bagAmount[1], bagAmount[2]})
		}
		for _, match := range bagAmounts {
			name := string(match[2])
			colorAncestors[name] = append(colorAncestors[name], color)
		}
	}

	// in how many different bags we can put a shiny gold bag
	stack := []string{"shiny gold"}
	ancestors := make(map[string]bool)
	for len(stack) > 0 {
		color := stack[0]
		stack = stack[1:]
		for _, ancestor := range colorAncestors[color] {
			ancestors[ancestor] = true
			stack = append(stack, ancestor)
		}
	}
	fmt.Println(len(ancestors))

	// how many other bags fit in a shiny gold bag
	fmt.Println(countBagsInside("shiny gold") - 1)
}
